### ISSUE TITLE
seccomp filter should return ENOSYS for unknown syscalls

### ISSUE THREAD

**Issue by jethrogb** [opencontainers/runc#2151]
Currently, the seccomp filter installed on Linux returns EPERM even for system calls that are unknown. This is [problematic when new system calls are added](https://github.com/rust-lang/rust/issues/65662) by Linux. Programs wishing to use the new system call will try to call it, and will implement a fallback mechanism when ENOSYS is returned (indicating the kernel doesn't support the call). However, when using containers, it will likely receive EPERM instead, failing instead of trying the fallback path.

In addition to the list of acceptable syscalls, the container definition should include a maximum known syscall number. The seccomp filter should be configured such that calls above the maximum return ENOSYS. When new syscalls are added, the maximum can be increased after the seccomp policy is updated.

**Comment by eero-t** [comment]
The inherent fragility of seccomp(): https://lwn.net/Articles/738694/

The discussion on the article is very instructive.  Basically whatever you do with seccomp, there are potential future landmines.  And those article comments didn't even go into kernel syscall & libc differences between HW architectures.

**Comment by richfelker** [comment]
`EPERM` just should not be used here. For a very large number of syscalls, `EPERM` has a well-defined meaning the application/library code making the call knows how to react to, and "blocked by container policy" is usually not in the scope of that meaning. For specific syscalls where the container policy does match the specified `EPERM` failure case, `EPERM` could reasonably be kept, but the default for all blocked syscalls should be `ENOSYS`.

**Comment by richfelker** [comment]
Note that this is breaking running of musl 1.2.0+ binaries - see https://github.com/docker/for-win/issues/8326 for another example - due to inability to perform the correct meaningful fallback. We can't just treat `EPERM` as cause for fallback too, since `EPERM` is also a meaningful error (and fallback to an old time32 syscall when seeing it would have TOCTOU bugs if the cause of `EPERM` went away between the two calls, making a race window where the Y2038-unsafe syscall gets used).

**Comment by cyphar** [comment]
On the other hand, if we return `ENOSYS` for something like `clone(CLONE_NEWUSER)` then `glibc` may get very confused ("clone isn't supported? oh i guess we're on Linux 1.2..."). IMHO we should split the current policy into "pretend the syscall doesn't exist" and "tell the program it's not allowed to do dodgy things". I agree that `EPERM` for unknown syscalls isn't ideal.

**Comment by richfelker** [comment]
I don't think glibc actually does that, but if you're blocking specific functionality of a syscall rather than the whole syscall, the choice of error code should be made on a per-syscall basis to match existing error semantics. For example `CLONE_NEWUSER` should do whatever the kernel does when the kernel is configured without user namespace support or with it restricted to use by root. But as-yet-unknown syscalls should all error with `ENOSYS`.

**Comment by fweimer** [comment]
@cyphar For unrecognized flag arguments, most system calls use an `EINVAL` error. `EPERM` is not correct there, either. (I think `clone` is an outlier and used to ignore invalid flags in the kernel, but that's been fixed on the kernel side, unlike `open`, unless you count `openat2`.)

**Comment by fweimer** [comment]
Note that this issue is now somewhat urgent because Linux has added an `faccessat2` system call, and many applications already use that already, by calling glibc's `faccessat` function.

**Comment by richfelker** [comment]
@fweimer Thanks for adding that. I really don't want this to remain a game of whack-a-mole where exceptions get added for each syscall once breakage is found. Upstream should do the right thing and stop producing `EPERM` entirely so fallbacks can work as intended.

**Comment by cyphar** [comment]
I want to point out that if you want to change the seccomp filter in Docker, you'll need to make an issue in the Docker repo. We don't control the seccomp filter that Docker uses. Until recently the runtime-spec didn't support custom return values, but that has changed so Docker will need to update their default seccomp profile.

**Comment by jethrogb** [comment]
@cyphar First, runc needs to add the ability to specify two different "defaults": one for known but not specifically specified syscalls, and one for unknown syscalls.

**Comment by richfelker** [comment]
OK, can someone familiar with Docker and how that works open an issue on their side and link to this one?

**Comment by thaJeztah** [comment]
the default seccomp profile in docker is in https://github.com/moby/moby/blob/master/profiles/seccomp/default_linux.go
a similar profile is also in containerd; https://github.com/containerd/containerd/blob/master/contrib/seccomp/seccomp_default.go

**Comment by cyphar** [comment]
@jethrogb I don't think runc is in a position to do that -- what set of syscalls are "known" is a property of the profile being written, not the container runtime. If you write a profile today and 50 syscalls get added next week, runc will know about those syscalls but the old profile will not.

Since Docker is the thing generating these profiles (and accepting user-specified profiles too), you would want the user-specified profile to say "this is the latest syscall at the time when I wrote this profile". The Docker-generated profile will then have a default action of `SECCOMP_RET_ERRNO` with errno `EPERM` (or whatever it is now) and then a rule will be added which checks whether the syscall number is larger than `$NEWEST_SYSCALL` and return `ENOSYS` in that case.

**Comment by jethrogb** [comment]
> I don't think runc is in a position to do that

Not today no, as you explain the profile description needs to be expanded to convey that information.

**Comment by thaJeztah** [comment]
Currently, runc also needs to know about the syscalls if they're included in the profile (https://github.com/opencontainers/runtime-spec/issues/1071), so if a profile specifies syscalls that runc doesn't know about, the container fails to start (see https://github.com/moby/moby/issues/41562).

**Comment by cyphar** [comment]
As a first-pass solution we can implement this in Docker et al by just assuming the largest syscall number specified in the profile at all is the last syscall before we give `ENOSYS`. This should work for most cases and we could extend the seccomp format in a separate PR.

@thaJeztah That is an issue but not really one that is super relevant here IMHO -- we just use libseccomp's syscall lookup features so updating libseccomp will update the supported syscall numbers. :shrug:

**Comment by cyphar** [comment]
Ah, I was wrong above -- I forgot that `runtime-spec` is entirely based around the syscall names so we'll need to modify the spec to work around this... Dammit.

**Comment by cyphar** [comment]
Ah, I was wrong above -- I forgot that `runtime-spec` is entirely based around the syscall names so if we want to do this nicely (with a rule which just does a `SCMP_CMP_GT` on the syscall number) we can't do it without changing the spec. That kinda sucks.

Docker could still work around it (by making the default `ENOSYS` and then having an explicit rule for all syscalls not accepted by other filters to return `EPERM`) but that's quite ugly...

**Comment by thaJeztah** [comment]
> is entirely based around the syscall names

Yes. That's what I tried to refer to in https://github.com/opencontainers/runc/issues/2151#issuecomment-728238461, but later saw my complete brain-fart that the issue I linked to was about capabilities, not syscalls 😂

**Comment by justincormack** [comment]
I talked about some of these issues in my Kubecon talk last week. In particular just listing calls to block, not an allowlist makes more sense, even if it results on failing open for new syscalls. Could tweak the error codes more easily in this case.

**Comment by thaJeztah** [comment]
> In particular just listing calls to block, not an allowlist makes more sense, even if it results on failing open for new syscalls

Would there be a risk if new syscalls are added that were not known at the time that the seccomp profile was generated? I think this thread (or the one on the mailing list) mentions the option to have the profile include information about the highest syscall number that was known at the time the profile was generated (potentially allowing new syscalls to be treated with some default (`ENOSYS` ?)); of course assuming that "new syscalls" would have higher numbers.

**Comment by richfelker** [comment]
@justincormack I think failing-closed is very reasonable here; it's just that the error code is wrong. There's nothing wrong with a fail-closed mechanism that effectively just emulates an old kernel, which is what you get if you fail with `ENOSYS`.

**Comment by cyphar** [comment]
It should remain an allow-list (fail-closed), the issue is that there are two kinds of failures that the current allow-list is handling with the same error code:

 1. Syscalls which we do not permit the container to run on purpose (such as `open_by_handle_at`). These must return `-EPERM` because `-ENOSYS` is simply the wrong error in that case. We explicitly are blocking these syscalls.
 2. Syscalls which are not included in the allow-list due to them not existing at the time the profile was created. `-ENOSYS` is the correct return value in this case.

The issue is that currently we pretend that all syscalls not included in the allow-list are in category (1) when in reality we should be defaulting to (2) for syscalls that were not known about at profile-creation time. In other words, the issue is not simply that "the error code is wrong" -- it's that there are two errors being handled with one error code. Changing the default action to return `-ENOSYS` would be just as incorrect at the current behaviour IMHO.

We could loosely infer which syscalls are in category (2) by assuming any syscall with a larger syscall number than the largest one in the profile is in category (2). However it would be nicer to have this behaviour be something that profile writers control (either by explicitly specifying the "largest known" syscall, or even better by allowing profiles to do `SCMP_CMP_` operations on raw syscall numbers -- which would also solve the issue @thaJeztah mentioned).

**Comment by richfelker** [comment]
I wouldn't characterize it as "just as incorrect". Semantically `ENOSYS` makes no sense for a syscall you're blocking for access control reasons, but at worst the cause gets misreported and the application possibly tries klunky fallbacks that will also fail due to being blocked. It's far less incorrect than the current situation.

Of course I'd like to see this solved in a way that distinguishes the two cases, if this can be done in a way that works right by default and doesn't depend on the profile author understanding why `EPERM` is wrong for unknown syscalls. If the behavior remains broken the way it is now unless everyone fixes their profiles, this is a major problem far worse than just changing the default could create.

**Comment by cyphar** [comment]
I guess that's a fair point. We could switch Docker to use `ENOSYS` as the default return value (to stop the bleeding) while we work on being able to nicely differentiate the two?

**Comment by fweimer** [comment]
>     2\. Syscalls which are not included in the allow-list due to them not existing at the time the profile was created.

@cyphar This rule worries me because it could mean that as soon as the profile is re-created with knowledge of the `faccessat2` system call (for example), the error code would turn from `ENOSYS` to `EPERM` because the system call transitions from unknown to implicitly filtered. I think we need a mechanism where simple rebuilds result in reproducible filters.

**Comment by jethrogb** [comment]
I'm not entirely sure, but I don't believe the seccomp profile is included in the Docker container image? This is just a runtime thing?

**Comment by fweimer** [comment]
I actually meant rebuilding the runtime against newer kernel headers/libseccomp. It should not have this effect, either.

**Comment by cyphar** [comment]
The default seccomp profile is not auto-generated -- it's [a hardcoded profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default_linux.go). Updating libseccomp doesn't have any effect on it. The only way what you describe could happen is if the built-in Docker profile was updated to include newer syscalls but one syscall was missed -- that would be a Docker bug and would be treated as one (it'd be like writing a profile which denied `open(2)` -- technically possible but unlikely to be missed).

**Comment by cyphar** [comment]
A few hitches in the above plan which I discovered while trying to implement this:

* There is no top-level `errnoRet` for the `defaultAction`. :facepalm: This does arguably suggest that technically the default errno is actually not defined by the specification.
* `libseccomp` doesn't support creating filters that discriminate by syscall number. :facepalm::facepalm: Seccomp itself definitely supports this (the syscall number is part of `seccomp_data` just like the arguments) but it seems like this would require us to generate our own BPF scripts to implement the behavior we want.

So it looks like the only way to stop the bleeding is to change the default errno to `-ENOSYS`. We could auto-generate the set of `-EPERM` rules based on which syscalls are present in the profile, but this would be probably be a bit too complicated to do correctly (you'd need to generate the inverse comparisons for every non-straightforward syscall).

/cc @opencontainers/runtime-spec-maintainers WDYT?

**Comment by giuseppe** [comment]
> So it looks like the only way to stop the bleeding is to change the default errno to `-ENOSYS` (which was the original suggestion -- sorry for not figuring this out earlier!). We could auto-generate the set of `-EPERM` rules based on which syscalls are present in the profile, but this would be probably be a bit too complicated to do correctly (you'd need to generate the inverse comparisons for every non-straightforward syscall).

Do you think it would be easier to enumerate all the syscalls not specified in the profile and let them fail with `-ENOSYS` so that the default is left unchanged (maybe adding a `errnoUnknownSyscall` knob in the spec)?  I think it simplifies handling the `-EPERM` case for the syscalls listed in the profile.  As long term fix, we can also try to fix it in libseccomp.  I don't see though, a way to retrieve `NR_syscalls` from userspace

**Comment by cyphar** [comment]
> Do you think it would be easier to enumerate all the syscalls not specified in the profile and let them fail with -ENOSYS

Did you mean `-EPERM`? Yes, we can do that but it'll increase the profile size and it's not as simple as that -- you'd need to generate inverse rules for `SCMP_CMP_*` comparison actions. Not impossible, just not ideal.

> I don't see though, a way to retrieve NR_syscalls from userspace

We wouldn't want to retrieve `NR_syscalls`, you'd just use the largest-numbered syscall in the profile. It's not ideal but there isn't a better solution without adding more information to profiles (which we will need to do eventually, but I'm talking about a hotfix here).

**Comment by giuseppe** [comment]
> > Do you think it would be easier to enumerate all the syscalls not specified in the profile and let them fail with -ENOSYS
> 
> Did you mean `-EPERM`? Yes, we can do that but it'll increase the profile size and it's not as simple as that -- you'd need to generate inverse rules for `SCMP_CMP_*` comparison actions. Not impossible, just not ideal.

no I meant `-ENOSYS`  :-)

After we generate all the rules as we do now, we can internally add the equivalent of (with the syscalls numerical values):
```
{
    "syscalls": [
        {
          "names": [
            "MAX_SYSCALL_KNOWN_IN_THE_PROFILE+1",
            "MAX_SYSCALL_KNOWN_IN_THE_PROFILE+2",
             ...
             "NR_syscalls"
          ],
          "action": "SCMP_ACT_ERRNO",
          "errnoRet": "38"
}
```
IMO, the advantage is that the default action (fail with `-EPERM`) will cover all the inverse rules so we don't have to worry about generating them.

**Comment by cyphar** [comment]
Not sure that's an improvement tbh. Also I just got linked to [a libseccomp thread](https://github.com/seccomp/libseccomp/issues/286) which covers what features we'd need from libseccomp.

**Comment by richfelker** [comment]
> Do you think it would be easier to enumerate all the syscalls not specified in the profile and let them fail with -ENOSYS so that the default is left unchanged (maybe adding a errnoUnknownSyscall knob in the spec)?

**NO!** The set you want to enumerate is unbounded/unknown. The whole problem we're trying to fix is that syscalls that aren't yet known are wrongly failing with `EPERM` when they should fail with `ENOSYS`. Just returning `ENOSYS` for syscall numbers that the *container runtime* knew about but that the profile author didn't know about doesn't help anything, because the apps run in the container could be using **any future syscall** that neither knows about, and those must all fail with `ENOSYS`.

**Comment by giuseppe** [comment]
> > 
> 
> **NO!** The set you want to enumerate is unbounded/unknown. The whole problem we're trying to fix is that syscalls that aren't yet known are wrongly failing with `EPERM` when they should fail with `ENOSYS`. Just returning `ENOSYS` for syscall numbers that the _container runtime_ knew about but that the profile author didn't know about doesn't help anything, because the apps run in the container could be using **any future syscall** that neither knows about, and those must all fail with `ENOSYS`.

I am aware of that issue.  That is why I was wondering if there is a way to retrieve `NR_syscalls` from user space at runtime (so that would be similar to what we do with capabilities), not when the container runtime is built and use numerical values for syscalls instead of names.  This would be a temporary solution until there is a proper fix in libseccomp.

My concern is that calculating the inverse rules for the seccomp profile can be difficult, e.g. we have some complex configuration for `socket(2)` for Podman: https://github.com/containers/common/blob/master/pkg/seccomp/seccomp.json#L790-L894

**Comment by fweimer** [comment]
>  That is why I was wondering if there is a way to retrieve `NR_syscalls` from user space at runtime (so that would be similar to what we do with capabilities), not when the container runtime is built and use numerical values for syscalls instead of names.

@giuseppe  This would still automatically switch from `ENOSYS` to `EPERM` once the host kernel supports the new system call, right? I don't think that's what we want.

**Comment by richfelker** [comment]
I'm not aware of any reliable way to do that at runtime, especially when the container runtime might already be running in a containerized environment where it can't access whatever `/proc` or `/sys` mechanism might let you discover the value or `NR_syscalls`, if one even exists.

I also don't see why you need to calculate any inverse rules. It seems you can just, for any syscall that appears in the profile, add a rule at the end to fail it with `EPERM`. If there are earlier allow rules, they'll take precedence. Then for any syscall that doesn't appear in the profile, you get the new default of `ENOSYS`.

As an aside, I'm also a bit horrified at the whole profile format. The DSL-in-JSON is not any simpler than just the raw BPF language, and in many ways harder to understand (what conditions are and vs or is not clear without having studied the DSL), and it doesn't even abstract things - I see a hard-coded 22 rather than `EINVAL`, making the whole thing of using syscall names rather than numbers to be more abstract and easier to port to other archs kinda pointless.

**Comment by giuseppe** [comment]
> @giuseppe This would still automatically switch from `ENOSYS` to `EPERM` once the host kernel supports the new system call, right? I don't think that's what we want.

what I am proposing is that the maximum syscall to consider is the one specified in the static seccomp profile, so that won't be affected by a newer runtime, libseccomp or kernel.  Anything that is below this maximum syscall will keep the existing semantic (by default fail with `EPERM`).  Anything that is above the maximum syscall in the profile, thus there are no rules specified (either pure or conditionals) can safely fail with `ENOSYS`.

The reason for detecting the maximum syscall available in the kernel at runtime is to enumerate the syscalls above the maximum syscall in the profile, and add a rule to let them fail with `ENOSYS`.

> I also don't see why you need to calculate any inverse rules. It seems you can just, for any syscall that appears in the profile, add a rule at the end to fail it with `EPERM`. If there are earlier allow rules, they'll take precedence. Then for any syscall that doesn't appear in the profile, you get the new default of `ENOSYS`.

How can we do it with libseccomp?  I was surprised to find out that a catch-all `seccomp_rule_add(3)` basically overrides any existing rule for that syscall.  From the man page:
```
All of the filter rules supplied by the calling application are combined into a union,
with additional  logic to eliminate redundant syscall filters.  For example, if a rule
is added which allows a given syscall with a specific set of argument values and later
a rule is added which allows the same syscall regardless the  argument  values  then
the first, more specific rule, is effectively dropped from the filter by the second more
generic rule.
```

> As an aside, I'm also a bit horrified at the whole profile format. The DSL-in-JSON is not any simpler than just the raw BPF language, and in many ways harder to understand (what conditions are and vs or is not clear without having studied the DSL), and it doesn't even abstract things - I see a hard-coded 22 rather than `EINVAL`, making the whole thing of using syscall names rather than numbers to be more abstract and easier to port to other archs kinda pointless.

Agreed, it is terrible to work with...

**Comment by richfelker** [comment]
> How can we do it with libseccomp? I was surprised to find out that a catch-all seccomp_rule_add(3) basically overrides any existing rule for that syscall. From the man page: ...

Can this just be fixed by fixing libseccomp to offer an API to do the right thing - add a fallback rule that doesn't override more specific existing rules? It seems like the source of the whole problem is libseccomp's awfulness and having the spec/runtime designed around its misdesign rather than doing something reasonable. But it should be possible to patch it up so that it's non-world-breaking in the mean time until it can be replaced entirely with something that's not broken-by-design.

**Comment by cyphar** [comment]
> The reason for detecting the maximum syscall available in the kernel at runtime is to enumerate the syscalls above the maximum syscall in the profile, and add a rule to let them fail with ENOSYS.

This wouldn't work because a program may try to use a syscall which doesn't exist on the host kernel either (think modern glibc on a 3.12 kernel trying to use `faccessat2`). Your approach of doing explicit `ENOSYS`es will require you to do it for all possible future syscalls and personally I think that's more painful (and less efficient IMHO) than the alternative. Though you could argue that technically it would make all of our seccomp profiles execute in `O(1)` time. :wink: 

> I also don't see why you need to calculate any inverse rules. It seems you can just, for any syscall that appears in the profile, add a rule at the end to fail it with EPERM. If there are earlier allow rules, they'll take precedence. Then for any syscall that doesn't appear in the profile, you get the new default of ENOSYS.

The reason I said we need to create inverse rules was simply out of an abundance of caution (not because I was sure it was a concrete issue we'd hit) -- libseccomp restructures rules a fair bit when generating the BPF output (to make the filter more performant) so you can't be sure that the rules will be in the order you specify them in. **However** I would expect the union behaviour @giuseppe mentions to only happen if the action is the same -- otherwise that's a bug in libseccomp's merging behaviour. My hope is that libseccomp probably does (or at least, it should) have the behaviour that subsequent rules are always evaluated later. I'll do some experiments to see if that's the case.

I think the best long-term solution is the one proposed in the libseccomp issue I linked -- specify the current kernel version (at time of profile writing) when generating the seccomp filter and libseccomp will create `ENOSYS` rules for you. libseccomp has to track all new syscalls (and their numbers on different architectures) so this is something we should entirely delegate to them. This also deals with non-contiguous syscall number addition (something which you can't handle without tracking all syscall additions by kernel version).

Regarding the DSL, yeah it's not great but (like quite a few hairy parts of the spec) we mostly imported it from Docker. :man_shrugging:

**Comment by cyphar** [comment]
@richfelker After thinking about this comment a bit more, I wanted to bring something up.

> I wouldn't characterize it as "just as incorrect". Semantically `ENOSYS` makes no sense for a syscall you're blocking for access control reasons, but at worst the cause gets misreported and the application possibly tries klunky fallbacks that will also fail due to being blocked. It's far less incorrect than the current situation.

`EPERM` (and `EACCES`) is used for fallback behaviour as well (in fact runc uses it in quite a few places during rootless execution as well as a couple of my other projects). If you did a GitHub code search for `os.IsPermission` checks I'd expect you'd find a fair amount of fallback behaviour. I'm pretty sure they're heavily used in quite a few rootless containers tools actually.

`ENOSYS` fallbacks are strictly speaking more common because glibc does them, but after thinking about it for a bit I'm no longer convinced that switching to `ENOSYS` won't also cause regressions. But I will do some experiments to see if we can best-effort generate `EPERM` rules. As I mentioned earlier, the DSL cannot specify the default action's errno so I think we're well within our rights (at least to stop the bleeding) to auto-generate some of this stuff in runc. Here's hoping no one complains.

**Comment by jethrogb** [comment]
> specify the current kernel version (at time of profile writing)

As discussed on the seccomp GH issue, this won't work. You need to specify the maximum known syscall number instead.

**Comment by giuseppe** [comment]
> This wouldn't work because a program may try to use a syscall which doesn't exist on the host kernel either (think modern glibc on a 3.12 kernel trying to use `faccessat2`). Your approach of doing explicit `ENOSYS`es will require you to do it for all possible future syscalls and personally I think that's more painful (and less efficient IMHO) than the alternative -- not to mention that the maximum syscall number could easily be increased in the future. Though you could argue that technically it would make all of our seccomp profiles execute in `O(1)` time. wink
> 
> >

From my experience getting stuff in a libseccomp release takes quite some time.  We'd probably need a temporary solution (so `NR_syscalls` could probably just be hardcoded with some margin for the future) until it can be done in a better way.



> **However** I would expect the union behaviour @giuseppe mentions to only happen if the action is the same -- otherwise that's a bug in libseccomp's merging behaviour. My hope is that libseccomp probably does (or at least, it should) have the behaviour that subsequent rules are always evaluated later. I'll do some experiments to see if that's the case.

From what I could see, that is not the case :-(  If you have a generic rule for a syscall, then adding a more specific one fails with `EPERM`

**Comment by cyphar** [comment]
> From my experience getting stuff in a libseccomp release takes quite some time. We'd probably need a temporary solution (so NR_syscalls could probably just be hardcoded with some margin for the future) until it can be done in a better way.

When you say `NR_syscalls` do you mean the number of syscalls in the latest kernel release or the number of syscalls in the latest kernel release? I thought you meant the number of syscalls on the currently running kernel (which my comment was in response to).

Btw, libseccomp updates aren't an issue for the maximum kernel version idea. libseccomp would generate `SCMP_CMP_GE` checks so you don't need to enumerate every possible syscall -- you couldn't specify a newer kernel version than the one libseccomp supports but in that case you also couldn't specify the syscall name either so it's no worse than the current situation.

**Comment by cyphar** [comment]
@jethrogb 

> As discussed on the seccomp GH issue, this won't work. You need to specify the maximum known syscall number instead.

Are you referring to your recent comment in seccomp/libseccomp#11? I didn't see it mentioned anywhere else. I don't think RHEL backporting policy means that the idea isn't workable -- Red Hat can also patch their libseccomp, and quite a few projects (incorrectly) check kernel versions for feature detection rather than the try-and-fail approach that is the "correct style". Even in the worst case, you'll get `ENOSYS` for a syscall which didn't exist in the upstream kernel version, instead of `EPERM`. I don't think that's really that bad.

**Comment by cyphar** [comment]
@jethrogb 

> As discussed on the seccomp GH issue, this won't work.

(I think you're referring to your recent comment in seccomp/libseccomp#11? I couldn't find another reference to this discussion.)

I disagree. Red Hat can also patch their version of libseccomp, and even in the worst case a syscall (which was going to be denied anyway) gets `ENOSYS` instead of `EPERM`. I don't think that's really that bad.

>  You need to specify the maximum known syscall number instead.

This has its own issues because syscalls aren't always added in consecutive order -- most recently, `openat2` was added several kernel releases before `clone_range`. If we want the long-term solution to not be based on a loose heuristic, maximum syscall number isn't all that much better IMHO.

**Comment by giuseppe** [comment]
> @giuseppe
> 
> > From my experience getting stuff in a libseccomp release takes quite some time. We'd probably need a temporary solution (so NR_syscalls could probably just be hardcoded with some margin for the future) until it can be done in a better way.
> 
> When you say `NR_syscalls` do you mean the number of syscalls in the latest kernel release or the number of syscalls in the latest kernel release? I thought you meant the number of syscalls on the currently running kernel (which my comment was in response to).
> Btw, libseccomp updates aren't an issue for the maximum kernel version idea. libseccomp would generate `SCMP_CMP_GE` checks so you don't need to enumerate every possible syscall -- you couldn't specify a newer kernel version than the one libseccomp supports but in that case you also couldn't specify the syscall name either so it's no worse than the current situation.

I agree with you, the longer term solution should be something like `SCMP_CMP_GE` comparing the syscall number and the maximum syscall defined in the profile, and it must be addressed in libseccomp.

Sorry for the confusion about `NR_syscalls` wrt the temporary fix.  I always referred to `NR_syscalls` as the " the number of syscalls on the running kernel".  You raised a valid point that even if we had a way to retrieve `NR_syscalls` at runtime, it could not be enough as a process can use a syscall not supported by the kernel and it would get `EPERM` instead of `ENOSYS`, so if we go this path, we'd need to use `NR_syscalls + N` (so at this point given `N`, `NR_sycalls` could just be defined at build time).

**Comment by richfelker** [comment]
The idea of hacks with a "+ N" threshold is preposterous. There is no guarantee there won't be arch-specific new syscalls outside of such a range, and you'll hit all sorts of issues with old runtimes that haven't been upgraded in a long time running newer binaries. This whole problem has a dead simple fix. Just do it.

**Comment by richfelker** [comment]
To clarify, the value of `NR_syscalls` on the running kernel is **completely a red herring**. It has nothing to do with the problem whatsoever, which involves only the API level the profile author was aware of and the API level the guest application is using. Neither of these have anything whatsoever to do with the host kernel.

While the most dead simple variant of the "dead simple fix" is just changing `EPERM` to `ENOSYS`, and this is the **right thing** to do, there's also a dead simple fix that preserves the `EPERM` behavior for existing known syscalls it wasn't explicitly assigned to: patching libseccomp to emit the logic in its output (rather than having to worry about complex new representations in input) where it normally emits the fallback rule for `EPERM`. It wouldn't even have to be tunable yet; it could just use a hard cutoff of "available in Linux 3.0" or something dumb like that.

**Comment by thaJeztah** [comment]
(Mostly) peanut-gallery comment: so I'm not _exactly_ sure why `EPERM` was picked as a default at the time; I could try searching history if there has been some discussion about selecting `EPERM` instead of `ENOSYS`, but it could've just been that it felt the most logical to indicate "this was blocked". If that was the _right_ choice in hindsight; perhaps yes, perhaps not (I guess that's what this discussion is about).

big disclaimer: I see concerns about backward incompatibility if `EPERM` would be changed to `ENOSYS`, and I haven't dug into if special handling is currently used to account for seccomp blocking syscalls.

That said, `ENOSYS` _feels_ natural to me; if the original intent was to distinguish a syscall from being "not supported" or "blocked"; that seems somewhat more appropriate for "debugging" while developing a profile. At runtime; should a (container) process really _care_ ? In the end; the syscall can't be made for _whatever_ reason (could be blocked, could be not supported); if the most natural way to handle fallbacks is to look for `ENOSYS`, that _feels_ most natural to me.

On a side-note; this discussion doesn't seem to be specific to runc; should it be migrated to the runtime-spec repo?

**Comment by richfelker** [comment]
@thaJeztah I agree it's not runc-specific and the problem should be fixed in the spec, but I'm not sure what technical fallout of migrating the issue would be. (Will existing links from off-site be lost? etc.) So I'll leave that to folks more familiar with GH.

**Comment by cyphar** [comment]
@thaJeztah 

> At runtime; should a (container) process really care ?

Programs in containers have fallback code that depends on `EPERM` (just as we do have code which depends on `ENOSYS`). I don't agree that `ENOSYS` feels more natural given that the explicit goal is to enact a security policy rather than hiding the existence of syscalls, but I don't think it's super important which feels better. `EPERM` is causing issues, and switching to `ENOSYS` *will also cause issues* (though admittedly on a smaller scale).

> On a side-note; this discussion doesn't seem to be specific to runc; should it be migrated to the runtime-spec repo?

This will need a runtime-spec change eventually, but the hotfix is going to be implemented in runc anyway so this repo probably is fine.

@richfelker 

> While the most dead simple variant of the "dead simple fix" is just changing EPERM to ENOSYS, and this is the right thing to do.

It's the least wrong thing to do at the moment (given the absence of other hotfixable options), but I do want to point out that doesn't mean it's correct. Even in this project, `runc` has three fallback paths which check `os.IsPermission` (one of them vendored from `github.com/opencontainers/selinux`) and all three of them will be broken by switching to `ENOSYS`. This means nested rootless containers won't work under newer runc versions unless we now special-case `ENOSYS` -- which we probably will do, but there will be countless other projects which will be in the same position.

Not arguing against doing it, it's just not as obvious of a decision as you make it sound.

**Comment by richfelker** [comment]
@cyphar: Is there anything wrong with switching default to `ENOSYS` but adding explicit `EPERM` for syscalls the profile doesn't allow, but that already existed in some chosen (I suggested Linux 3.0 as a random choice but it really doesn't matter) baseline syscall API level? If this is hard to represent in the profile DSL, just have libseccomp replace the fallback rule with this logic, which should be nothing more than adding a few lines of static BPF appends to the libseccomp source.

**Comment by AkihiroSuda** [comment]
@cyphar Is there any action on plan? If not, can we remove this from rc93 milestone?

**Comment by richfelker** [comment]
Please don't let this get relegated to some indeterminate future. It's impacting users and pressuring glibc to make bad decisions about use of new syscalls, and should be very high priority for a fix.

**Comment by AkihiroSuda** [comment]
@richfelker Would you want to open a PR?

**Comment by richfelker** [comment]
Possibly, but I'm not the most qualified for working with the codebase and there should first be agreement on what policy is acceptable since it's pointless to implement a policy that won't be adopted. My previous comment https://github.com/opencontainers/runc/issues/2151#issuecomment-735962372 suggests what might be a low-impact albeit somewhat ugly fix.

**Comment by cyphar** [comment]
I will work in on this this week, and I agree this should be done ASAP to avoid making glibc work around us in an even less optimal way.

**Comment by giuseppe** [comment]
@cyphar are you fine if I open a PR for runtime-specs to allow overriding the default errno for the seccomp profile (similarly to https://github.com/opencontainers/runtime-spec/pull/1041)?

**Comment by giuseppe** [comment]
> @cyphar are you fine if I open a PR for runtime-specs to allow overriding the default errno for the seccomp profile (similarly to [opencontainers/runtime-spec#1041](https://github.com/opencontainers/runtime-spec/pull/1041))?

I've opened it here: https://github.com/opencontainers/runtime-spec/pull/1087

**Comment by kallisti5** [comment]
What *is* the actual fix for this one?  I think the massive number of references above show this is breaking a *LOT* of stuff in a lot of places.  A lot of the bugs above mention using older os images as the "workaround", but nobody has found a solution.

In the concourse bug I opened above, 32-bit binaries are failing to run in runc containers.  statx is allowed via the seccomp config, but is failing with EPERM for an unknown reason.

I opened https://github.com/opencontainers/runtime-spec/issues/1122 to "change the default errnoRet", but even that seems like an incorrect solution.

Does *anyone* have an idea on what is going on?

**Comment by cyphar** [comment]
> In the concourse bug I opened above, 32-bit binaries are failing to run in runc containers. statx is allowed via the seccomp config, but is failing with EPERM for an unknown reason.

**Comment by cyphar** [comment]
> In the concourse bug I opened above, 32-bit binaries are failing to run in runc containers. statx is allowed via the seccomp config, but is failing with EPERM for an unknown reason.

That sounds like a bug in our BPF patching code (that really shouldn't be happening -- we do have handling for different architectures including the 32-on-64-bit "architecture" so it's a bit puzzling that it's not working as expected). I will take a look at this this week.

**Comment by kallisti5** [comment]
@cyphar Thanks :-)    Would moving away from Fedora 34 to something like Alpine be a temporary workaround?   We really didn't run into this until we upgraded the container host from Fedora 33 to 34.

**Comment by cyphar** [comment]
I'm not sure to be honest -- I'm also confused how the container host upgrade could've caused this as well. If there's an issue with our 32-bit compat handling it should happen on every system. The obvious contender (kernel version change) doesn't really explain it either.

**Comment by kallisti5** [comment]
@cyphar   @fweimer-rh keeps mentioning something changed in Fedora 34, but really hasn't identified what beyond this:

> As I said on the Fedora bug, it's a bug in the container environment. Fedora 34 simply uses slightly different parts of the Linux system call interface than previous versions. 

@fweimer , @fweimer-rh , can you provide clarification on what changed in Fedora 34?

**Comment by cyphar** [comment]
The ENOSYS boundary is primarily set by the seccomp profile not libseccomp (if libseccomp doesn't know about a syscall it's ignored, but it's still primarily set by the profile -- which wouldn't have changed between versions since it's a Docker-version thing not related to the rest of the host and statx in particular has been allowed for quite a while). But yeah, as I said I'll take a look at this this wee.
